api/opensource/mfx_dispatch/linux/mfxloader.cpp

```diff
@@ -0,0 +1,583 @@
+{
+    { eMFXInit, "MFXInit", VERSION(1, 0) },
+    { eMFXInitEx, "MFXInitEx", VERSION(1, 14) },
+    { eMFXClose, "MFXClose", VERSION(1, 0) },
+    { eMFXJoinSession, "MFXJoinSession", VERSION(1, 1) },
+#include "mfxvideo_functions.h"
+    { eNoMoreFunctions }
+};
+
+typedef mfxStatus (MFX_CDECL *CreatePluginPtr)(mfxPluginUID, mfxPlugin*);
+
+class LoaderCtx;
+
+class PluginCtx
+{
+public:
+  PluginCtx(LoaderCtx& loader)
+    : m_loader(loader)
+  {}
+
+  mfxStatus Load(const mfxPluginUID& uid, mfxU32 version, const char *path);
+  mfxStatus Unload();
+
+  inline mfxPluginUID getUID() const { return m_uid; }
+
+private:
+  LoaderCtx& m_loader;
+  std::shared_ptr<void> m_dlh;
+  CreatePluginPtr m_create_plugin = nullptr;
+  mfxPluginUID m_uid{};
+  mfxPlugin m_plugin{};
+  mfxPluginParam m_plugin_param{};
+};
+
+class LoaderCtx
+{
+public:
+  mfxStatus Init(mfxInitParam& par);
+  mfxStatus Close();
+
+  mfxStatus LoadPlugin(const mfxPluginUID& uid, mfxU32 version, const char *path);
+  mfxStatus UnloadPlugin(const mfxPluginUID& uid);
+
+  inline void* getFunction(Function func) const {
+    return m_table[func];
+  }
+
+  inline mfxSession getSession() const {
+    return m_session;
+  }
+
+  inline mfxIMPL getImpl() const {
+    return m_implementation;
+  }
+
+  inline mfxVersion getVersion() const {
+    return m_version;
+  }
+
+private:
+  std::shared_ptr<void> m_dlh;
+  mfxVersion m_version{};
+  mfxIMPL m_implementation{};
+  mfxSession m_session = nullptr;
+  void* m_table[eFunctionsNum]{};
+
+  std::mutex m_guard;
+  std::list<PluginCtx> m_plugins;
+};
+
+struct GlobalCtx
+{
+  std::mutex m_mutex;
+  std::list<PluginInfo> m_plugins;
+};
+
+static GlobalCtx g_GlobalCtx;
+
+std::shared_ptr<void> make_dlopen(const char* filename, int flags)
+{
+  return std::shared_ptr<void>(
+    dlopen(filename, flags),
+    [] (void* handle) { if (handle) dlclose(handle); });
+}
+
+mfxStatus LoaderCtx::Init(mfxInitParam& par)
+{
+  if (!MFX_IMPL_BASETYPE(par.Implementation) &&
+      !(par.Implementation & MFX_IMPL_VIA_VAAPI)) {
+    return MFX_ERR_UNSUPPORTED;
+  }
+  if (par.Implementation & MFX_IMPL_AUDIO) {
+    return MFX_ERR_UNSUPPORTED;
+  }
+
+  std::vector<std::string> libs;
+
+  if (MFX_IMPL_BASETYPE(par.Implementation) == MFX_IMPL_AUTO ||
+      MFX_IMPL_BASETYPE(par.Implementation) == MFX_IMPL_AUTO_ANY) {
+    libs.emplace_back(LIBMFXHW);
+    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXHW);
+    libs.emplace_back(LIBMFXSW);
+    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXSW);
+  } else if (par.Implementation & MFX_IMPL_HARDWARE ||
+             par.Implementation & MFX_IMPL_HARDWARE_ANY) {
+    libs.emplace_back(LIBMFXHW);
+    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXHW);
+  } else if (par.Implementation & MFX_IMPL_SOFTWARE) {
+    libs.emplace_back(LIBMFXSW);
+    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXSW);
+  } else {
+    return MFX_ERR_UNSUPPORTED;
+  }
+
+  mfxStatus mfx_res = MFX_ERR_UNSUPPORTED;
+
+  for (auto lib: libs) {
```

Comment:
these strings may be quite large, so I think better to use references here:

for (auto& libs: libs)